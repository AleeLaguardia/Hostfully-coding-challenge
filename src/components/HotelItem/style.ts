import styled from "styled-components";

export const Container = styled.div`
  display: flex;
  min-height: 260px;
  height: 260px;
  width: 100%;
  border-radius: 15px;
  border: 1px solid ${(props) => props.theme.colors.alto};
  padding: 12px 10px;
  margin: 10px 0;
  transition: 0.15s;
  cursor: pointer;

  @media (max-width: 1000px) {
    min-height: 160px;
    height: 160px;
  }

  @media (max-width: 600px) {
    min-height: 130px;
    height: 130px;
  }
`;

export const Image = styled.img`
  height: 236px;
  width: 236px;
  object-fit: cover;
  border-radius: 10px;
  transition: 0.15s;

  @media (max-width: 1000px) {
    height: 136px;
    width: 136px;
  }

  @media (max-width: 600px) {
    height: 106px;
    width: 106px;
  }
`;

export const InfoContainer = styled.div`
  display: flex;
  flex-direction: column;
  height: 100%;
  margin: 0 10px;
  width: 50%;
  justify-content: space-between;

  .name-container {
    display: flex;
    flex-direction: column;
    height: fit-content;
    width: 100%;
  }

  .name {
    font-size: ${(props) => props.theme.sizes.xxxl};
    color: ${(props) => props.theme.colors.purple};
    font-family: ${(props) => props.theme.fonts.poppins};
    font-weight: 500;

    @media (max-width: 600px) {
      font-size: ${(props) => props.theme.sizes.m};
    }
  }

  .description {
    font-size: ${(props) => props.theme.sizes.xs};
    color: ${(props) => props.theme.colors.emperor};
    font-family: ${(props) => props.theme.fonts.poppins};
    font-weight: 500;
    height: 55px;
    overflow: hidden;
    display: -webkit-box;
    -webkit-box-orient: vertical;
    -webkit-line-clamp: 3;
    text-overflow: ellipsis;

    @media (max-width: 600px) {
    height: 35px;
    -webkit-line-clamp: 2;
    }
  }

  .address {
    font-size: ${(props) => props.theme.sizes.xs};
    color: ${(props) => props.theme.colors.lynch};
    font-family: ${(props) => props.theme.fonts.poppins};
    font-weight: 500;  
  }
`;

export const RatingContainer = styled.div`
  display: flex;
  flex: 1;
  flex-direction: column;
  justify-content: space-between;
  align-items: flex-end;
  padding: 0 10px;


  @media (max-width: 700px) {
    padding: 0 0 0 10px;    
  }
`;

export const Rating = styled.div`
  display: flex;
  height: fit-content;
  width: fit-content;
  align-items: center;

  .price-container {
    display: flex;
    flex-direction: column;
    align-items: flex-end;
    width: fit-content;
    height: fit-content;
    background-color: ${(props) => props.theme.colors.purple};
    padding: 5px;
    border-radius: 10px;
    padding-right: 10px;

    @media (max-width: 600px) {
      display: none;
    }
  }

  span {
    font-family: ${(props) => props.theme.fonts.poppins};
    font-weight: 500;
    color: ${(props) => props.theme.colors.lynch};
    margin-left: 5px;
  }

  .rate {
    font-size: ${(props) => props.theme.sizes.m};
  }

  .price {
    font-size: ${(props) => props.theme.sizes.s};
    color: ${(props) => props.theme.colors.white};
    font-weight: 400;

    @media (max-width: 700px) {
      font-size: ${(props) => props.theme.sizes.xxs};
    }
  }
`