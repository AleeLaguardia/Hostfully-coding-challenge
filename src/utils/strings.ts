export const PATH = {
  HOME: '/',
  BOOKING: '/booking',
  RESERVATION: '/reservation',
};

export const HOME = {
  TITLE_FORM: "Book your stay right now!",
  TITLE_PIC: "Book Your Stay",
  ADULTS: "How many adults",
  CHILDREN: "How many children",
}

export const BOOKING = {
  NO_DATA: "No hotels found in that area",
  ALSO_TRY: "But you can also try:"
};