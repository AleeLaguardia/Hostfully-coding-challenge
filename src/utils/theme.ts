import { DefaultTheme } from "styled-components";

import {
  white,
  alabaster,
  emperor,
  galery,
  gray,
  black,
  purple,
  purpleHover,
  darkPurple,
  radicalRed,
  lightPurple,
  silverChalice,
  alto,
  lynch,
  selectiveYellow,
  chelseaGem,
} from './styling/colors';
import {
  poppins,
  playwrite,
} from './styling/fonts';
import {
  xs,
  xxs,
  xxxs,
  s,
  m,
  l,
  xl,
  xxxl,
} from './styling/sizes';

export const theme: DefaultTheme = {
  colors: {
    white,
    alabaster,
    emperor,
    galery,
    gray,
    black,
    purple,
    purpleHover,
    darkPurple,
    radicalRed,
    lightPurple,
    silverChalice,
    alto,
    lynch,
    selectiveYellow,
    chelseaGem,
  },
  fonts: {
    poppins,
    playwrite,
  },
  sizes: {
    xs,
    xxs,
    xxxs,
    s,
    m,
    l,
    xl,
    xxxl,
  },
}