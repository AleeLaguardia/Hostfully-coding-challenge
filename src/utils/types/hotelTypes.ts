import { LatLngExpression } from "leaflet";

export type Location = {
  coordinates: LatLngExpression;
  type: string;
}

export type Room = {
  BaseRate: number;
  BedOptions?: string;
  Description?: string;
  Description_fr?: string;
  SleepsCount?: number
  SmokingAllowed?: boolean;
  Tags?: string[];
  Type?: string;
}

export type AddressType = {
  StreetAddress: string;
  City: string;
  StateProvince: string;
  PostalCode: string;
  Country: string;
}

export type Hotel = {
  HotelId: string;
  ImageSource: string;
  HotelName: string;
  Description: string;
  Description_fr: string;
  Category: string;
  Tags: string[];
  ParkingIncluded: boolean,
  LastRenovationDate: string;
  Rating: number;
  Address: AddressType;
  Rooms: Room[];
  Location: Location;
}