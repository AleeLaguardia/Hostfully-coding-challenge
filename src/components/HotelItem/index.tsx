import React from "react";
import { Container, Image, InfoContainer, Rating, RatingContainer } from "./style";
import StarIcon from '../../assets/icons/star.svg';
import { Hotel } from "../../utils/types/hotelTypes";

interface Props {
  hotel: Hotel;
  onClick: () => void;
  totalPrice: string;
}

const HotelItem: React.FC<Props> = ({ hotel, onClick, totalPrice }) => {
  const { StreetAddress, City, StateProvince } = hotel.Address;
  
  return (
    <Container onClick={onClick} key={hotel.HotelId}>
      <Image draggable={false} src={hotel?.ImageSource || ''} alt="source-image" />
      <InfoContainer>
        <div className="name-container">
          <span className="name">{hotel.HotelName}</span>
          <span className="description">{hotel.Description}</span>
        </div>
        <span className="address">{StreetAddress}, {City}, {StateProvince}</span>
      </InfoContainer>
      <RatingContainer>
        <Rating>
          <img src={StarIcon} alt="StarIcon" />
          <span className="rate">{hotel.Rating}</span>
        </Rating>
        <Rating>
          <div className="price-container">
            <span className="price">$ {hotel.Rooms[0].BaseRate} night</span>
            <span className="price">{totalPrice} total</span>
          </div>
        </Rating>
      </RatingContainer>
    </Container>
  );
};

export default HotelItem