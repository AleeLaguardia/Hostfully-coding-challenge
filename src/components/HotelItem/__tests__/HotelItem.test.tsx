import React from 'react';
import configureStore from 'redux-mock-store';
import { render, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import HotelItem from '..';
import { Hotel } from '../../../utils/types/hotelTypes';
import { BrowserRouter } from 'react-router-dom';
import { ThemeProvider } from 'styled-components';
import { theme } from '../../../utils/theme';
import { Provider } from 'react-redux';

const mockHotel: Hotel = {
  HotelId: '1',
  HotelName: 'Test Hotel',
  Description: 'This is a test description for the test hotel.',
  ImageSource: 'https://example.com/image.jpg',
  Address: {
    StreetAddress: '123 Test St',
    City: 'Test City',
    StateProvince: 'TS',
    Country: 'USA',
    PostalCode: '000000',
  },
  Category: 'Test Category',
  Location: {
    coordinates: [1, 1],
    type: 'test type'
  },
  Description_fr: 'Description test',
  LastRenovationDate: 'Renovation Test',
  ParkingIncluded: false,
  Rating: 5.0,
  Tags: [],
  Rooms: [
    {
      BaseRate: 100,
    },
  ],
};

const mockStore = configureStore([]);

describe('HotelItem Component', () => {
  let store: any;
  const onClickMock = jest.fn();

  beforeEach(() => {
    store = mockStore({
      user: {
        destination: 'Paris',
        adults: '2 adults',
        children: '1 child',
        date: [new Date('2024-07-14'), new Date('2024-07-17')],
      },
    });
  });

  it('renders hotel item with default style', () => {
    const { asFragment } = render(
      <BrowserRouter>
        <ThemeProvider theme={theme}>
          <Provider store={store}>
            <HotelItem
              hotel={mockHotel}
              onClick={onClickMock}
              totalPrice="$300.00"
            />
          </Provider>
        </ThemeProvider>
      </BrowserRouter>
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('renders hotel name, description, and address', () => {
    const { getByText, getByAltText } = render(
      <BrowserRouter>
        <ThemeProvider theme={theme}>
          <Provider store={store}>
            <HotelItem
              hotel={mockHotel}
              onClick={onClickMock}
              totalPrice="$300.00"
            />
          </Provider>
        </ThemeProvider>
      </BrowserRouter>
    );

    expect(getByText('Test Hotel')).toBeInTheDocument();
    expect(getByText('This is a test description for the test hotel.')).toBeInTheDocument();
    expect(getByText('123 Test St, Test City, TS')).toBeInTheDocument();

    const image = getByAltText('source-image') as HTMLImageElement;
    expect(image).toBeInTheDocument();
    expect(image.src).toBe('https://example.com/image.jpg');
  });

  it('renders default image when ImageSource is not provided', () => {
    const { getByAltText } = render(
      <BrowserRouter>
        <ThemeProvider theme={theme}>
          <Provider store={store}>
            <HotelItem
              hotel={{ ...mockHotel, ImageSource: '' }}
              onClick={onClickMock}
              totalPrice="$300.00"
            />
          </Provider>
        </ThemeProvider>
      </BrowserRouter>
    );

    const image = getByAltText('source-image') as HTMLImageElement;
    expect(image).toBeInTheDocument();
  });

  it('handles click event', () => {
    const { getByText } = render(
      <BrowserRouter>
        <ThemeProvider theme={theme}>
          <Provider store={store}>
            <HotelItem
              hotel={mockHotel}
              onClick={onClickMock}
              totalPrice="$300.00"
            />
          </Provider>
        </ThemeProvider>
      </BrowserRouter>
    );

    fireEvent.click(getByText('Test Hotel'));
    expect(onClickMock).toHaveBeenCalled();
  });

  it('renders hotel rating and price', () => {
    const { getByText, getByAltText } = render(
      <BrowserRouter>
        <ThemeProvider theme={theme}>
          <Provider store={store}>
            <HotelItem
              hotel={mockHotel}
              onClick={onClickMock}
              totalPrice="$300.00"
            />
          </Provider>
        </ThemeProvider>
      </BrowserRouter>
    );

    const starIcon = getByAltText('StarIcon') as HTMLImageElement;
    expect(starIcon).toBeInTheDocument();
    expect(getByText('5')).toBeInTheDocument();
    expect(getByText('$ 100 night')).toBeInTheDocument();
    expect(getByText('$300.00 total')).toBeInTheDocument();
  });
});
