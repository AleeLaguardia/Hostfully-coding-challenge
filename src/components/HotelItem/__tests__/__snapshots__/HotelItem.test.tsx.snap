// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HotelItem Component renders hotel item with default style 1`] = `
<DocumentFragment>
  <div
    class="sc-aYaIB fOHEvt"
  >
    <img
      alt="source-image"
      class="sc-gEvDqW gNMiKU"
      draggable="false"
      src="https://example.com/image.jpg"
    />
    <div
      class="sc-eqUzNf kmvpQh"
    >
      <div
        class="name-container"
      >
        <span
          class="name"
        >
          Test Hotel
        </span>
        <span
          class="description"
        >
          This is a test description for the test hotel.
        </span>
      </div>
      <span
        class="address"
      >
        123 Test St, Test City, TS
      </span>
    </div>
    <div
      class="sc-fqkwJk YOsBz"
    >
      <div
        class="sc-dcJtft hOHCai"
      >
        <img
          alt="StarIcon"
          src="star.svg"
        />
        <span
          class="rate"
        >
          5
        </span>
      </div>
      <div
        class="sc-dcJtft hOHCai"
      >
        <div
          class="price-container"
        >
          <span
            class="price"
          >
            $ 100 night
          </span>
          <span
            class="price"
          >
            $300.00 total
          </span>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
